tidy3d/plugins/microwave/array_factor.py

```diff
@@ -761,28 +877,496 @@ def array_factor(
+        discriminator=TYPE_TAG_STR,
+    )
+
+    window_z: Optional[RectangularWindowType] = pd.Field(
+        None,
+        title="Z Axis Window",
+        description="Window type used to taper array antenna along z axis.",
+        discriminator=TYPE_TAG_STR,
+    )
+
+    # validate that at least one window is provided
+    @classmethod
+    def isotropic_window(cls, window: RectangularWindowType) -> RectangularTaper:
+        """
+        Set the same window along x, y, and z dimensions.
+
+        Parameters
+        ----------
+        window: RectangularWindowType
+            A supported 1D window type from ``RectangularWindowType``.
+
+        Returns
+        -------
+        RectangularTaper
+            A ``RectangularTaper`` instance with all three dimensions set to the specified window.
+        """
+        return cls(window_x=window, window_y=window, window_z=window)
+
+    @pd.root_validator
+    def check_at_least_one_window(cls, values):
+        if not any([values.get("window_x"), values.get("window_y"), values.get("window_z")]):
+            raise ValueError("At least one window (x, y, or z) must be provided.")
+        return values
+
+    def amp_multipliers(
+        self, array_size: tuple[PositiveInt, PositiveInt, PositiveInt]
+    ) -> tuple[ArrayLike, ArrayLike, ArrayLike]:
+        """
+        Method ``amp_multipliers()`` computes rectangular taper amplitude for phased array antennas.
+
+        Parameters:
+        ----------
+        array_size: tuple[PositiveInt, PositiveInt, PositiveInt]
+            A tuple of array size along x,y, and z axes.
+
+        Returns:
+        --------
+        tuple[ArrayLike, ArrayLike, ArrayLike]
+            a tuple of three 1D numpy arrays with taper amplitudes along x,y, and z axes.
+        """
+
+        effective_size = tuple(dim if dim is not None else 1 for dim in array_size)
+
+        amps = (
+            window.get_weights(effective_size[ind])
+            if window is not None
+            else np.ones(effective_size[ind])
+            for ind, window in enumerate([self.window_x, self.window_y, self.window_z])
         )
 
-        # Calculate the overall array factor
-        array_factor = af_x * af_y * af_z
+        return amps
 
-        return array_factor
+
+class RadialTaper(AbstractTaper):
+    """Class for Radial Taper."""
+
+    window: TaylorWindow = pd.Field(
+        ..., title="Window Object", description="Window type used to taper array antenna."
+    )
+
+    # Change TaylorWindow mode to call _TaylorRadial
+    @pd.root_validator(pre=False)
+    def force_radial_taylor(cls, values):
+        if not values:
+            return values
+
+        w = values.get("window")
+
+        # Allow TaylorWindow instance; reject anything else early
+        if not isinstance(w, TaylorWindow):
+            raise TypeError(
+                f"Expected a TaylorWindow instance for 'window', got {type(w).__name__}."
+            )
```

Comment:
i think we don't need to validate this, as it should be validated already in the field's type declaration